exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -448,39 +427,44 @@ public Builder withValidators(short validatorCount) {
     }
 
     /**
-     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     * Adds a service artifact which would be deployed by the TestKit. Several service artifacts
+     * can be added.
+     *
+     * <p>Note that the corresponding service instance with equal serviceArtifactId should be
+     * created with {@link #withService(ServiceArtifactId, String, int, Any)}.
      */
-    public Builder withService(Class<? extends ServiceModule> serviceModule) {
-      services.add(serviceModule);
+    public Builder withDeployedService(ServiceArtifactId serviceArtifactId, String artifactFilename) {
+      serviceArtifactFilenames.put(serviceArtifactId, artifactFilename);
       return this;
     }
 
     /**
-     * Adds services with which the TestKit would be instantiated.
-     */
-    @SafeVarargs
-    public final Builder withServices(Class<? extends ServiceModule> serviceModule,
-                                      Class<? extends ServiceModule>... serviceModules) {
-      return withServices(Lists.asList(serviceModule, serviceModules));
-    }
-
-    /**
-     * Adds services with which the TestKit would be instantiated.
+     * Adds a service specification with which the TestKit would create the corresponding service
+     * instance. Several service specifications can be added.
+     *
+     * <p>Note that the corresponding service artifact with equal serviceArtifactId should be
+     * deployed with {@link #withDeployedService(ServiceArtifactId, String)}.
      */
-    public Builder withServices(Iterable<Class<? extends ServiceModule>> serviceModules) {
-      Iterables.addAll(services, serviceModules);
+    public Builder withService(ServiceArtifactId serviceArtifactId,
+                               String serviceName,
+                               int serviceId,
+                               Any configuration) {
+      ServiceSpec serviceSpec =
+          ServiceSpec.newInstance(serviceName, serviceId, configuration.toByteArray());
+      services.put(serviceArtifactId, serviceSpec);
```

Comment:
Shan't we check here that the corresponding artifact is already added? I.e.,
shall we enforce the rule 'an artifact first, its instances — next' which doesn't sound
too restrictive? Or allow any order?

Reply:
I don't think we need such a restriction, we will check that all deployed artifacts match the instances when we build the TestKit. While 'an artifact first, its instances - next' sounds intuitive to me too and follows the correct order, I don't see anything wrong with the reverse order.
Plus it seems weird when `builder.a().b().build()` works, but `builder.b().a().build()` - doesn't.

Reply:
In usual operation, if you add a service, an artifact must be deployed as its dependency first. If we enforce the rule, we provide earlier error reporting and force the client code to perform the operations in the same order in which they normally (and subsequently) happen; but we might prevent some exotic use cases. I wonder why it might have to be done in the opposite order?

Reply:
Ok, we can add this restriction. Then we don't need to check that any services were created, but not deployed when we build TestKit.